core/trino-main/src/main/java/io/trino/metadata/AbstractPropertyManager.java

```diff
@@ -69,28 +75,191 @@ protected final void doRemoveProperties(K propertiesKey)
+        // nonNullPropertiesBuilder, otherwise put its name into nullPropertyNamesBuilder
+        for (String name : separatedProperties.defaultPropertyNames) {
+            PropertyMetadata<?> propertyMetadata = supportedPropertiesMetadata.get(name);
+            if (propertyMetadata == null) {
+                throw new TrinoException(
+                        propertyError,
+                        format("%s does not support %s property '%s'",
+                                formatPropertiesKeyForMessage(catalogNameForDiagnostics, propertiesKey),
+                                propertyType,
+                                name));
+            }
+            Object value = propertyMetadata.getDefaultValue();
+            if (value != null) {
+                nonNullPropertiesBuilder.put(name, value);
+            }
+            else {
+                nullPropertyNamesBuilder.add(name);
+            }
+        }
+
+        return new Properties(nonNullPropertiesBuilder.build(), nullPropertyNamesBuilder.build());
+    }
+
     protected final Map<String, Object> doGetProperties(
             K propertiesKey,
             String catalogNameForDiagnostics,
-            Map<String, Expression> sqlPropertyValues,
+            Iterable<Property> properties,
             Session session,
             Metadata metadata,
             AccessControl accessControl,
-            Map<NodeRef<Parameter>, Expression> parameters,
-            boolean setDefaultProperties)
+            Map<NodeRef<Parameter>, Expression> parameters)
     {
-        Map<String, PropertyMetadata<?>> supportedProperties = connectorProperties.get(propertiesKey);
-        if (supportedProperties == null) {
-            throw new TrinoException(NOT_FOUND, formatPropertiesKeyForMessage(catalogNameForDiagnostics, propertiesKey) + " not found");
+        return doGetProperties(
+                propertiesKey,
+                catalogNameForDiagnostics,
+                separateProperties(propertiesKey, catalogNameForDiagnostics, properties).nonDefaultProperties,
+                session,
+                metadata,
+                accessControl,
+                parameters);
+    }
+
+    /**
+     * Separates the {@code Property}'s in {@code properties} based on whether they are set to DEFAULT or not
+     */
+    private SeparatedProperties separateProperties(
+            K propertiesKey,
+            String catalogNameForDiagnostics,
+            Iterable<Property> properties)
+    {
+        Map<String, Expression> nonDefaultProperties = new HashMap<>();
+        Set<String> defaultPropertyNames = new HashSet<>();
+        for (Property property : properties) {
+            String name = property.getName().getValue().toLowerCase(ENGLISH);   // property names are case-insensitive and normalized to lower case
+            if (nonDefaultProperties.containsKey(name) || defaultPropertyNames.contains(name)) {
+                throw new TrinoException(
```

Comment:
Could we throw semantic exception here pointing at proper property?

Reply:
In this file, `TrinoException`'s are always constructed using one of `TrinoException`'s constructors - they are never constructed using the `semanticException` helper methods. I followed the existing style in this file.